crates/telio-firewall/src/firewall.rs

```diff
@@ -1401,7 +1501,195 @@ pub mod tests {
+
+        const V4_SRC_IP: &str = "8.8.8.8";
+        const V4_DST_IP: &str = "127.0.0.1";
+
+        const V6_SRC_IP: &str = "2001:4860:4860::8888";
+        const V6_DST_IP: &str = "::1";
+
+        for (i, request) in V4_REQUEST_TYPES.iter().enumerate() {
+            for (j, reply) in V4_REPLY_TYPES.iter().enumerate() {
+                let fw = StatefullFirewall::new();
+                let outbound = make_packet(&make_icmp4, V4_SRC_IP, V4_DST_IP, (*request).into(), 1, 1);
+                let inbound = make_packet(&make_icmp4, V4_DST_IP, V4_SRC_IP, (*reply).into(), 1, 1);
+                assert_eq!(fw.process_outbound_packet(&make_peer(), &outbound), true);
+                assert_eq!(fw.process_inbound_packet(&make_peer(), &inbound), i == j);
+            }
+            for (j, reply) in V6_REPLY_TYPES.iter().enumerate() {
+                let fw = StatefullFirewall::new();
+                let outbound = make_packet(&make_icmp4, V4_SRC_IP, V4_DST_IP, (*request).into(), 1, 1);
+                let inbound = make_packet(&make_icmp6, V6_DST_IP, V6_SRC_IP, (*reply).into(), 1, 1);
+                assert_eq!(fw.process_outbound_packet(&make_peer(), &outbound), true);
+                assert_eq!(fw.process_inbound_packet(&make_peer(), &inbound), false);
+            }
+        }
+
+        for (i, request) in V4_REPLY_TYPES.iter().enumerate() {
+            for (j, reply) in V4_REPLY_TYPES.iter().enumerate() {
+                let fw = StatefullFirewall::new();
+                let outbound = make_packet(&make_icmp4, V4_SRC_IP, V4_DST_IP, (*request).into(), 1, 1);
+                let inbound = make_packet(&make_icmp4, V4_DST_IP, V4_SRC_IP, (*reply).into(), 1, 1);
+                assert_eq!(fw.process_outbound_packet(&make_peer(), &outbound), true);
+                assert_eq!(fw.process_inbound_packet(&make_peer(), &inbound), false);
+            }
+            for (j, reply) in V6_REPLY_TYPES.iter().enumerate() {
+                let fw = StatefullFirewall::new();
+                let outbound = make_packet(&make_icmp4, V4_SRC_IP, V4_DST_IP, (*request).into(), 1, 1);
+                let inbound = make_packet(&make_icmp6, V6_DST_IP, V6_SRC_IP, (*reply).into(), 1, 1);
+                assert_eq!(fw.process_outbound_packet(&make_peer(), &outbound), true);
+                assert_eq!(fw.process_inbound_packet(&make_peer(), &inbound), false);
+            }
+        }
+
+        for (i, request) in V6_REQUEST_TYPES.iter().enumerate() {
+            for (j, reply) in V6_REPLY_TYPES.iter().enumerate() {
+                let fw = StatefullFirewall::new();
+                let outbound = make_packet(&make_icmp6, V6_SRC_IP, V6_DST_IP, (*request).into(), 1, 1);
+                let inbound = make_packet(&make_icmp6, V6_DST_IP, V6_SRC_IP, (*reply).into(), 1, 1);
+                assert_eq!(fw.process_outbound_packet(&make_peer(), &outbound), true);
+                assert_eq!(fw.process_inbound_packet(&make_peer(), &inbound), i == 0 && j == 0);
```

Comment:
Why `i == 0 && j == 0`? Advert should also be a valid response for solicit

Reply:
I think it should be `i == j` as for v4 messages. If it's not then there is a bug somewhere. It might be that you overwrite some bytes in `make_packet` function for those messages even though they do not have id and seq fields.

Reply:
The complicated thing here, which I haven't really thought about how to approach is that `BLOCKED_TYPES` for IPv6 contains `RouterAdvert`, `NeighborSolicit` and `NeighborAdvert`. Since those types are in `BLOCKED_TYPES` I added them to the cache handling like the other types, but they are essentially unusable given that they're in `BLOCKED_TYPES` so it would probably make more sense to remove them from the cache handling completely. That would reject any inbound ICMP packet with those type, but then again, that is already happening

Reply:
I removed handling of those ICMP types from the cache since they will be blocked anyways, and I simplified this test